diff --git a/src/daemon/custom/huawei/hisi-hdmi/pulseaudio.conf b/src/daemon/custom/huawei/hisi-hdmi/pulseaudio.conf
new file mode 100644
index 0000000..0ed7727
--- /dev/null
+++ b/src/daemon/custom/huawei/hisi-hdmi/pulseaudio.conf
@@ -0,0 +1,9 @@
+# Add an incremental coverage profile for Huawei platforms.
+
+default-sample-format = s16le
+default-sample-rate = 48000
+default-sample-channels = 2
+default-channel-map = front-left,front-right
+
+default-fragments = 2
+default-fragment-size-msec = 20
